zkevm-circuits/src/pi_circuit.rs

```diff
@@ -1021,88 +1572,130 @@ impl<F: Field> PiCircuitConfig<F> {
         // offset += 1;
 
         // block state root
-        let state_root = rlc(extra.state_root.to_fixed_bytes(), randomness);
-        let state_root_cell = region.assign_advice(
-            || "state.root",
-            self.raw_public_inputs,
-            offset,
-            || Value::known(state_root),
+        self.assign_raw_bytes(
+            region,
+            &extra.state_root.to_fixed_bytes(),
+            rpi_bytes_keccakrlc,
+            rpi_bytes,
+            rpi_bytes_length_acc,
+            challenges,
         )?;
-        raw_pi_vals[offset] = state_root;
-        offset += 1;
 
         // previous block state root
-        let prev_state_root = rlc(extra.prev_state_root.to_fixed_bytes(), randomness);
-        let prev_state_root_cell = region.assign_advice(
-            || "parent_block.hash",
-            self.raw_public_inputs,
-            offset,
-            || Value::known(prev_state_root),
+        self.assign_raw_bytes(
+            region,
+            &extra.prev_state_root.to_fixed_bytes(),
+            rpi_bytes_keccakrlc,
+            rpi_bytes,
+            rpi_bytes_length_acc,
+            challenges,
         )?;
-        raw_pi_vals[offset] = prev_state_root;
-        Ok([state_root_cell, prev_state_root_cell])
+
+        Ok(())
     }
 
-    /// Assign `rpi_rlc_acc` and `rand_rpi` columns
+    /// Assign digest related columns
     #[allow(clippy::type_complexity)]
-    fn assign_rlc_pi(
+    fn assign_rpi_digest(
         &self,
         region: &mut Region<'_, F>,
-        rand_rpi: F,
-        raw_pi_vals: Vec<F>,
-    ) -> Result<(AssignedCell<F, F>, AssignedCell<F, F>), Error> {
+        digest: [u8; 32],
+        challenges: &Challenges<Value<F>>,
+    ) -> Result<(AssignedCell<F, F>, AssignedCell<F, F>, AssignedCell<F, F>), Error> {
+        assert!(digest.len() == N_BYTES_WORD);
+
         let circuit_len = self.circuit_len();
-        assert_eq!(circuit_len, raw_pi_vals.len());
+        let challenge = challenges.evm_word();
+        let mut cells = vec![];
+        digest
+            .to_vec()
+            .iter()
+            .chain(
+                (0..circuit_len - 1 - digest.len()).map(|_| &0u8), // padding to max circuit length
```

Comment:
perhaps we don't really need to pad it to `circuit_len`? since I found after offset `32` everything is assigned to be `F::zero()`, is that true?

Reply:
right, and this part removed in new bottom up computation style